### PR TITLE
*: make analyze buckets number configurable

### PR DESCRIPTION
**Pull request by lamxTyler**
### What problem does this PR solve? 
Make analyze buckets number configurable.

### What is changed and how it works?
Extend the syntax of analyze, so we can use statement like `analyze table t with 4 buckets` to configure the bucket number.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - None

Related changes

 - Need to update the documentation

PTAL @coocood @zz-jason @winoros

### CODE REVIEW

File: statistics/update_test.go [hunk 1]
```diff
@@ -890,7 +886,7 @@ func (s *testStatsUpdateSuite) TestLogDetailedInfo(c *C) {
 	for i := 0; i < 20; i++ {
 		testKit.MustExec(fmt.Sprintf("insert into t values (%d, %d, %d)", i, i, i))
 	}
-	testKit.MustExec("analyze table t")
+	testKit.MustExec("analyze table t limit 4 buckets")
```
Comment (shenli) [inline statistics/update_test.go]: Why use 4 buckets here?

File: parser/parser.y [hunk 2]
```diff
@@ -1225,14 +1227,23 @@ TableToTable:
...
+|	"ANALYZE" "TABLE" TableName "INDEX" IndexNameList MaxNumBuckets
+	{
+		$$ = &ast.AnalyzeTableStmt{TableNames: []*ast.TableName{$3.(*ast.TableName)}, IndexNames: $5.([]model.CIStr), IndexFlag: true, MaxNumBuckets: $6.(uint64)}
+	}
+
+MaxNumBuckets:
+	{
+		$$ = uint64(0)
+	}
+|	"LIMIT" NUM "BUCKETS"
```
Comment (zz-jason) [inline parser/parser.y]: how about change the syntax to `analyze table xx with xx buckets`?

maybe we can lean from mysql: https://dev.mysql.com/doc/refman/8.0/en/analyze-table.html

File: executor/builder.go [hunk 3]
```diff
@@ -1385,10 +1387,14 @@ func (b *executorBuilder) buildAnalyze(v *plan.Analyze) Executor {
 		baseExecutor: newBaseExecutor(b.ctx, v.Schema(), v.ExplainID()),
 		tasks:        make([]*analyzeTask, 0, len(v.ColTasks)+len(v.IdxTasks)),
 	}
+	maxNumBuckets := v.MaxNumBuckets
+	if maxNumBuckets == 0 {
```
Comment (zz-jason) [inline executor/builder.go]: IMHO, it's better to more this validation to the parser or planner.

File: statistics/update_test.go (hunk 1 above)
Comment (lamxTyler) [reply statistics/update_test.go]: It is the test's requirement. Each bucket should contain more than 1 values.

**Comment by lamxTyler** [comment]
@shenli It depends on the user. Maybe the user wants to limit the stats memory usage, or just want to try and see if the plan gets better. Anyway, it is optional.

**Comment by coocood** [comment]
@lamxTyler 
Seems the bucket number is not persistent, will the buckets gets splitted or merged by feedback?

File: plan/planbuilder.go [hunk 4]
```diff
@@ -680,7 +680,12 @@ func (b *planBuilder) buildAnalyzeAllIndex(as *ast.AnalyzeTableStmt) Plan {
 	return p
 }
 
+const defaultNumBuckets = 256
```
Comment (winoros) [inline plan/planbuilder.go]: defaultMax?

**Comment by coocood** [comment]
LGTM

File: plan/planbuilder.go [hunk 5]
```diff
@@ -680,7 +680,12 @@ func (b *planBuilder) buildAnalyzeAllIndex(as *ast.AnalyzeTableStmt) Plan {
 	return p
 }
 
+const defaultMaxNumBuckets = 256
+
 func (b *planBuilder) buildAnalyze(as *ast.AnalyzeTableStmt) (Plan, error) {
+	if as.MaxNumBuckets == 0 {
+		as.MaxNumBuckets = defaultMaxNumBuckets
```
Comment (zz-jason) [inline plan/planbuilder.go]: should we set an upper limit for the `MaxNumBuckets`

**Pull request merged**